### ISSUE TITLE
install error under jruby 1.7.18

### ISSUE THREAD

**Issue by jjb** [thoughtbot/capybara-webkit#725]
I've successfully installed other versions of capybara-webkit with other versions of jruby on the same machine. I'm not sure exactly what combinations are problematic.

Let me know what other info you would like and I'll provide right away.

```
➔ gem install capybara-webkit -v '1.4.1' ; alert
Building native extensions.  This could take a while...
ERROR:  Error installing capybara-webkit:
	ERROR: Failed to build gem native extension.

    /Users/john/.rbenv/versions/jruby-1.7.18/bin/jruby extconf.rb
NotImplementedError: C extension support is not enabled. Pass -Xcext.enabled=true to JRuby or set JRUBY_OPTS.

   (root) at /Users/john/.rbenv/versions/jruby-1.7.18/lib/ruby/shared/mkmf.rb:8
  require at org/jruby/RubyKernel.java:1071
   (root) at /Users/john/.rbenv/versions/jruby-1.7.18/lib/ruby/shared/rubygems/core_ext/kernel_require.rb:1
   (root) at extconf.rb:1


Gem files will remain installed in /Users/john/.rbenv/versions/jruby-1.7.18/lib/ruby/gems/shared/gems/capybara-webkit-1.4.1 for inspection.
Results logged to /Users/john/.rbenv/versions/jruby-1.7.18/lib/ruby/gems/shared/gems/capybara-webkit-1.4.1/./gem_make.out
```

**Comment by mhoran** [comment]
It looks like this is due to a dependency on `mkmf` introduced in 07976a2d78940ad3f62c5ae1be737199c528067b. 

While capybara-webkit is not a C extension, we do utilize the C extension compilation framework to compile the `webkit_server`. You'll need to follow the instructions from the error, e.g. set `cext.enabled=true` in `.jrubyrc`. `.jrubyrc` needs to be in your home directory.

**Comment by jferris** [comment]
@mhoran it seems like this isn't a bug. If that's correct, should this be closed?

**Comment by mhoran** [comment]
Agreed. We could probably remove the `mkmf` dependency with some effort, but the workaround should work just fine.

**Comment by mico** [comment]
I got this error, after enabled C extension:
    
    Installing capybara-webkit 1.6.0 with native extensions

    Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    /Users/mico/.rbenv/versions/jruby-1.7.21/bin/jruby -r ./siteconf20150816-92068-cpid4l.rb extconf.rb
    /Users/mico/.rbenv/versions/jruby-1.7.21/lib/ruby/shared/mkmf.rb:14: Use RbConfig instead of obsolete and deprecated Config.
    mkmf.rb can't find header files for ruby at /Users/mico/.rbenv/versions/jruby-1.7.21/lib/native/include/ruby/ruby.h

    extconf failed, uncaught signal 1

**Comment by jjb** [comment]
I get the same error as @mico. I use rbenv and I think it does not distribute the header file. Looks like C extension support is pretty neglected in JRuby: https://github.com/jruby/jruby/wiki/C-Extension-Alternatives

**Comment by jferris** [comment]
We aren't actually writing a C extension. We have C++ code which gets compiled into a separate binary at the same time you install the Gem.

We're currently hooking into the Rubygems hooks for C extensions in order to trigger the compile step. If somebody has ideas for an alternative, I'd be open to them, and it may eliminate the issues with JRuby.

**Comment by synth** [comment]
Has anyone else had problems with Travis-CI, JRuby, and Capybara-webkit? I have a project that I just upgraded to latest capybara-webkit(1.10.1), and now Travis-CI won't build the project because its saying it has disabled C Extensions. Is there a workaround for this? I've reached out to Travis-CI, but they are notoriously slow with responding...

**Comment by BanzaiMan** [comment]
I looked at this issue at @synth's urging, and I was able to convince JRuby 1.7.19 to install capybara-webkit version 1.10.1 in the context of Travis CI. I am sure you can adjust the process in other situations.
​
See https://github.com/BanzaiMan/travis_production_test/commit/f6f255c8fe0d8f184321d1f2b84358218f529bea
​
## Steps

​A few things needs to happen in order for you to make this happen:

1. ​Tweak mkmf.rb so that the absence of ruby.h does not abort gem installation. https://github.com/BanzaiMan/travis_production_test/blob/f6f255c8fe0d8f184321d1f2b84358218f529bea/.travis.yml#L18
1. You will need `qmake` command. This is not available on our build images, so you can either run `sudo apt-get install qt4-qmake` for Ubuntu or use the APT addon (https://github.com/BanzaiMan/travis_production_test/blob/f6f255c8fe0d8f184321d1f2b84358218f529bea/.travis.yml#L9). I'm sure you can find a suitable package in Homebrew on OS X.
1. Ensure that `-Xcext.enabled=true` is set: https://github.com/BanzaiMan/travis_production_test/blob/f6f255c8fe0d8f184321d1f2b84358218f529bea/.travis.yml#L13
1. Ensure that you specify `mime-types` less than 3.0: https://github.com/BanzaiMan/travis_production_test/blob/f6f255c8fe0d8f184321d1f2b84358218f529bea/Gemfile#L4. `mime-types` 3.0 depends on mime-types-data, which requires Ruby 2.0. While JRuby 1.9 offers some preliminary 2.x-compatibility, it is not something I recommend for production use.

​You can see how all of this work together in https://travis-ci.org/BanzaiMan/travis_production_test/builds/124626897.

I looked at JRuby's packaging, but couldn't quite figure out why `ruby.h` is not there any more (even if it is, I think the path to it is incorrect). Overall, though, JRuby's C extension support (and by extension `mkmf` support) is dead, so I would not fret too much about it.

## Disclaimer
Notice that I've only tested installation. `capybara-webkit` thus installed on JRuby may or may not function as intended.

This process is admittedly brittle, and not clear how much longer will be valid.

**Please caution when you do use this.**

**Comment by jferris** [comment]
Thanks for putting all this together.

It's worth noting that we don't actually build C extensions. Rather, we use the C extension hook in RubyGems to build an external binary that we then communicate with over pipes and network sockets. I don't believe `ruby.h` or anything similar is necessary; we require the ability to build C++ applications using Qt, which actually has nothing to do with Ruby or C extensions. Maybe there's a better hook we could use in general?

**Comment by headius** [comment]
FWIW, jruby/jruby#4031 has been merged in for JRuby 9.1.3.0, so non-cext extconf builds will work ok.

**Comment by jjb** [comment]
😻
